#2 feature(lab2): Реализовал 2 лабораторную работу

inittose opened the pull request:
- Добавил перечисление Category и внедрил в класс Item
- Добавил класс Address, написал логику к нему и внедрил в проект
- Добавил класс Store, переписал сериализацию и внедрил в проект
- Также переписал ItemFactory и CustomerFactory для корректной генерации

inittose committed:
feature(lab2): Добавление категорий товара

- В класс Item добавлено свойство Category
- Во вкладке ItemsTab добавлен ComboBox для категорий
- Реализована логика работы с  ComboBox

inittose committed:
feature(lab2): Добавил класс Address

- Создал AddressControl, логика прописана (еще не проверена)

inittose committed:
feature(lab2): AddressControl внедрен в ItemsTab

- Строковый адрес в Customer заменен на Address
- Проверена логика работы AddressControl

inittose committed:
feature(lab2): Добавил и внедрил класс Store

- Сериализация работает с одним классом Store
- MainForm управляет данными класса Store
- Также поправил CustomerFactory, теперь он генерирует адреса для Address

inittose committed:
docs(lab2): UML-диаграмма 2 лабораторной

- Также добавил выводи диаграммы в README

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Address.cs:
```diff
@@ -0,0 +1,181 @@
+        /// <param name="street">Улица.</param>
+        /// <param name="building">Номер здания.</param>
+        /// <param name="apartment">Номер квартиры/помещения.</param>
+        public Address(int index, string country, string city, 
```
Если выполняешь перенос аргументов, то нужно, чтобы каждый аргумент находился на своей строке (первый аргумент тоже переносится). При этом все аргументы отстоят на одну табуляцию от объявления метода (но это тебе IDE сделает сама, скорее всего)

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Store.cs:
```diff
@@ -0,0 +1,32 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Лишние using. Во всем проекте

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -14,6 +14,16 @@ public static class CustomerFactory
         private const int MAX_ROWS = 1000;
 
+        /// <summary>
+        /// Минимальное значение почтового индекса
```
Точка в конце предложения. Снизу тоже

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -31,10 +41,24 @@ public static Customer GetRandomCustomer()
+        {
+            var random = new Random();
+            var RandomPostIndex = random.Next(MIN_POST_INDEX, MAX_POST_INDEX);
+            return RandomPostIndex;
```
Можно сократить и сразу вернуть random.Next....

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/ItemFactory.cs:
```diff
@@ -33,8 +33,9 @@ public static Item GetRandomItem()
+            var randomCategory = (Category)random.Next(0, Enum.GetNames(typeof(Category)).Length);
 
-            var randomItem = new Item(itemName, itemInfo, itemCost);
+            var randomItem = new Item(itemName, itemInfo, itemCost, randomCategory);
```
здесь тоже можно сократить и сразу вернуть new Itew(...)

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Controls/AddressControl.cs:
```diff
@@ -0,0 +1,314 @@
+        /// <summary>
+        /// Цвет <see cref="TextBox"/>, успешно прошедшего валидацию. 
+        /// </summary>
+        private Color RightInputColor { get; } = Color.White;
```
В ItemTab ты тоже используешь такие переменные (нарушаешь принцип Don’t repeat yourself), сделай их static readonly и вынеси в static класс, который можешь назвать, типо AppColor

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/ItemsTab.cs:
```diff
@@ -22,21 +43,27 @@ public partial class ItemsTab : UserControl
-        private List<Item> Items { get; } = Serializer.GetItems();
-
         /// <summary>
         /// Инициализировать компонент, 
```
пиши комментарии в стиле "что делает" - "инициализирует"

inittose committed:
fix(lab2): Исправлен баг с CustomerTab + правки

- Убрал лишние using в файлах
- Создал статический класс AppColors для цветов в программе
- Подредактировал комментарии

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/AppColors.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace ObjectOrientedPractics.Services
+{
+    public static class AppColors
```
Комментарий для класса

inittose committed:
docs(lab2): Добавил AppColors UML-диаграмму

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/Serializer.cs:
```diff
@@ -36,68 +30,37 @@ static Serializer()
 
         /// <summary>
-        /// Сериализовать данные о покупателях и сохранить изменения.
+        /// Сериализовать данные о товарах и покупателях и сохранить изменения.
```
сериализует и сохраняет, используй вопрос "что делает?" когда пишешь комментарии

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/Serializer.cs:
```diff
@@ -36,68 +30,37 @@ static Serializer()
 
         /// <summary>
-        /// Десериализовать данные о покупателях.
+        /// Десериализовать данные о товарах и покупателях.
```
десериализует

inittose committed:
style(lab2): Правки по комментариям

- Пример: Сериализовать -> Сериализует (в методах)

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -31,7 +31,7 @@ public static class CustomerFactory
 
         /// <summary>
-        /// Получить экзепляр класса <see cref="Customer"/> с начальными случайными значениями.
+        /// Получает экзепляр класса <see cref="Customer"/> с начальными случайными значениями.
```
не получает, а возвращает

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -50,7 +50,7 @@ public static Customer GetRandomCustomer()
 
         /// <summary>
-        /// Получить случайный почтовый индекс.
+        /// Получает случайный почтовый индекс.
```
Возвращает

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/ItemFactory.cs:
```diff
@@ -21,7 +21,7 @@ public static class ItemFactory
 
         /// <summary>
-        /// Получить экзепляр класса <see cref="Item"/> с начальными случайными значениями.
+        /// Получает экзепляр класса <see cref="Item"/> с начальными случайными значениями.
```
Возвращает

inittose committed:
docs(lab2): Микро-правки в XML-комментариях

- Получает -> Возвращает

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -14,14 +14,24 @@ public static class CustomerFactory
         /// Путь до базы данных покупателей.
         /// </summary>
         private static string FileName { get; } =
             $@"{Directory.GetCurrentDirectory()}\..\..\Services\Databases\CustomersData.txt";
```
Вместо абсолютных путей используй относительные. Для этого используй две точки "../"вместо {Directory.GetCurrentDirectory()}. Твой подход будет тоже работать, как и относительные, но я всё равно рекомендую использовать относительные, потому что они выглядят лаконичнее

inittose committed:
style(lab2): Перенес аргументы в методах

- Перешел на относительный путь
